src/sentry/utils/redis.py

```diff
@@ -98,6 +102,110 @@ def execute_command(self, *args, **kwargs):
             return super(self.__class__, self).execute_command(*args, **kwargs)
 
 
+class FailoverRedis(StrictRedis):
+    """
+    Single host redis client implementation with retry logic intended to
+    survive failover events. Retry logic uses capped exponential backoff with
+    jitter.
+
+    https://redis.io/commands/failover
+
+    Failover sequence:
+
+    1. The primary will internally start a CLIENT PAUSE WRITE, which will pause
+    incoming writes and prevent the accumulation of new data in the replication
+    stream. From this point all writes to the primary instance fails with
+    ReadOnlyError.
+
+    2. The primary will monitor its replicas, waiting for a replica to indicate
+    that it has fully consumed the replication stream. If the primary has
+    multiple replicas, it will only wait for the first replica to catch up.
+
+    3. The primary will then demote itself to a replica. This is done to
+    prevent any dual master scenarios.
+
+    4. The previous primary will send a special PSYNC request to the target
+    replica, PSYNC FAILOVER, instructing the target replica to become a
+    primary.
+
+    5. Once the previous primary receives acknowledgement the PSYNC FAILOVER
+    was accepted it will unpause its clients.
+
+    In addition, the Memorystore for Redis, which is the main target of this implementation states:
+
+    When the primary node fails over to the replica, existing connections to
+    the primary endpoint of the instance are dropped. The instance is
+    unavailable for a few seconds while the new primary reconnects. On
+    reconnect, your application is automatically redirected to the new primary
+    node using the same connection string or IP address. You do not need to
+    update your application after a failover.
+
+    https://cloud.google.com/memorystore/docs/redis/high-availability#how_a_failover_affects_your_application
+
+    """
+
+    def __init__(
+        self,
+        *args,
+        _retries: int = 10,
+        _backoff_min: float = 0.2,
+        _backoff_max: int = 5,
+        _backoff_multiplier: float = 2,
```

Comment:
These default values were chosen by testing this implementation against memorystore, triggering the manual failover. (https://cloud.google.com/memorystore/docs/redis/initiating-manual-failover)

During the testing, with `socket_timeout=1`, the client would recover after 4-7 retries.